### PR TITLE
TimestepArrays

### PR DESCRIPTION
**Pull request by ckingdon95**

### CODE REVIEW

**Comment by ckingdon95** [comment]
need to merge fund.jl PR to fix parameter dictionary type, then update the dependency link to fund, then travis tests should pass and this is ready to merge.

File: src/Mimi.jl [hunk 1]
```diff
@@ -7,18 +7,21 @@ using NamedArrays
 
 export
     ComponentState, run_timestep, run, @defcomp, Model, setindex, addcomponent, setparameter,
-    connectparameter, setleftoverparameters, getvariable, adder, MarginalModel, getindex,
-    getdataframe, components, variables, getvpd, unitcheck, set_external_parameter, plot, getindexcount,
-    getindexvalues, getindexlabels, delete!, get_unconnected_parameters, Timestep, isfirsttimestep, isfinaltimestep
+    connectparameter, setleftoverparameters, getvariable, adder, MarginalModel, ConnectorComp, getindex,
+    getdataframe, components, variables, getvpd, unitcheck, plot, getindexcount,
+    getindexvalues, getindexlabels, delete!, get_unconnected_parameters, Timestep, isfirsttimestep,
+    isfinaltimestep, TimestepVector, TimestepMatrix, hasvalue
```
Comment (davidanthoff) [inline src/Mimi.jl]: Do we need to export ``Timestep``, ``TimestepVector`` and ``TimestepMatrix``? Users shouldn't really construct these types themselves, right?

File: src/mimi-core.jl [hunk 2]
```diff
@@ -115,6 +117,20 @@ function variables(mi::ModelInstance, componentname::Symbol)
     return fieldnames(mi.components[componentname].Variables)
 end
 
+function isuniform(values::Vector)
```
Comment (davidanthoff) [inline src/mimi-core.jl]: Could you add a comment above this that describes what this does? One short sentence is enough.

File: src/mimi-core.jl [hunk 3]
```diff
@@ -682,14 +757,41 @@ function build(m::Model)
...
-        comp = t(m.numberType, m.indices_counts)
+        ext_connections = filter(x->x.component_name==c.name, m.external_parameter_connections)
+        # ext_params = map(x->x.param_name, ext_connections)
+        ext_params = Dict(x.param_name => x.external_parameter for x in ext_connections)
+
+        int_connections = filter(x->x.target_component_name==c.name, m.internal_parameter_connections)
+        int_params = Dict(x.target_parameter_name => x for x in int_connections)
+
+        constructor = Expr(:call, c.component_type, m.numberType, :(Val{$(c.offset)}), :(Val{$duration}), :(Val{$(c.final)}))
+        for (pname, p) in get_parameters(m, c)
```
Comment (davidanthoff) [inline src/mimi-core.jl]: Could you add some comments here? For example: "The following loop does X"

File: src/mimi-core.jl [hunk 4]
```diff
@@ -707,6 +809,8 @@ function build(m::Model)
         param = x.external_parameter
         if isa(param, ScalarModelParameter)
             setfield!(builtComponents[x.component_name].Parameters, x.param_name, param.value)
+        # elseif isa(getfield(builtComponents[x.component_name].Parameters, x.param_name), Array)
+        #     setfield!(builtComponents[x.component_name].Parameters, x.param_name, param.values.data)
```
Comment (davidanthoff) [inline src/mimi-core.jl]: Delete?

File: src/mimi-core.jl [hunk 5]
```diff
@@ -773,10 +881,10 @@ function run(mi::ModelInstance, ntimesteps, indices_values)
     while !finished(clock)
         for (i, (name, c)) in enumerate(components)
             if gettime(clock) >= offsets[i] && gettime(clock) <= final_times[i]
+                # println("running component $name in $(gettime(clock))")
```
Comment (davidanthoff) [inline src/mimi-core.jl]: Delete

**Pull request merged**